couchbase.go

```diff
@@ -0,0 +1,308 @@
+package couchbase
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/couchbase/gocb/v2"
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/database/dbplugin"
+	"github.com/hashicorp/vault/sdk/database/helper/credsutil"
+	"github.com/hashicorp/vault/sdk/database/helper/dbutil"
+)
+
+var _ = fmt.Printf
+
+const (
+	couchbaseTypeName        = "couchbase"
+	defaultCouchbaseUserRole = `{"Roles": [{"role":"ro_admin"}]}`
+)
+
+var (
+	_ dbplugin.Database = &CouchbaseDB{}
+)
+
+type RolesAndGroups struct {
+	Roles   []gocb.Role `json:"roles"`
+	Groups   []string `json:"groups"`
+}
+
+// New implements builtinplugins.BuiltinFactory
+func New() (interface{}, error) {
+	db := new()
+	// Wrap the plugin with middleware to sanitize errors
+	dbType := dbplugin.NewDatabaseErrorSanitizerMiddleware(db, db.secretValues)
+	return dbType, nil
+}
+
+func new() *CouchbaseDB {
+	connProducer := &couchbaseDBConnectionProducer{}
+	connProducer.Type = couchbaseTypeName
+
+	credsProducer := &credsutil.SQLCredentialsProducer{
+		DisplayNameLen: 50,
+		RoleNameLen:    50,
+		UsernameLen:    50,
+		Separator:      "-",
+	}
+
+	db := &CouchbaseDB{
+		couchbaseDBConnectionProducer: connProducer,
+		CredentialsProducer:           credsProducer,
+	}
+
+	return db
+}
+
+// Run instantiates a CouchbaseDB object, and runs the RPC server for the plugin
+func Run(apiTLSConfig *api.TLSConfig) error {
+	dbType, err := New()
+	if err != nil {
+		return err
+	}
+
+	dbplugin.Serve(dbType.(dbplugin.Database), api.VaultPluginTLSProvider(apiTLSConfig))
+
+	return nil
+}
+
+type CouchbaseDB struct {
```

Comment:
All exported symbols should have a comment - it's fine for that comment to be very short if the symbol is simple/self-explanatory, although this particular one probably warrants a sentence or two.

I'll save making the same comment on other symbols, but there are a few other examples too.

Reply:
Done.